### PR TITLE
Redux recompose

### PR DESCRIPTION
**Pull request by andyOlarte514**
# Summary
- Integration with `redux-recompose`

# Trello card
- https://trello.com/c/DbWtknPz

### CODE REVIEW

File: training-react/tic-tac-toe/src/app/screens/Home/screens/Profile/components/Form/index.js [hunk 1]
```diff
@@ -23,19 +23,19 @@ function Form({ handleSubmit, msgUpdate, hasError }) {
...
         </button>
       </form>
-      {!hasError ? (
-        <span className={styles.successUpdate}>{msgUpdate}</span>
-      ) : (
-        <span className={styles.failureUpdate}>{msgUpdate}</span>
-      )}
+      {isSuccessUpdate ? (
+        <span className={styles.successUpdate}>Success update user data</span>
+      ) : hasError ? (
```
Comment (henryzarza) [inline training-react/tic-tac-toe/src/app/screens/Home/screens/Profile/components/Form/index.js]: You could replace this for: `hasError && <span className={styles.failureUpdate}>Error in update user data</span>`

File: training-react/tic-tac-toe/package.json [hunk 2]
```diff
@@ -6,13 +6,15 @@
     "@fortawesome/fontawesome-svg-core": "^1.2.6",
     "@fortawesome/free-solid-svg-icons": "^5.4.1",
     "@fortawesome/react-fontawesome": "^0.1.3",
+    "ajv": "^6.5.4",
```
Comment (henryzarza) [inline training-react/tic-tac-toe/package.json]: Do you need this?

File: training-react/tic-tac-toe/src/app/screens/Login/index.js [hunk 3]
```diff
@@ -5,19 +5,17 @@ import PropTypes from 'prop-types';
...
-  return <Layout onSubmit={handleSubmit} hasError={hasErrorAuth} msgError={msgError} />;
+function Login({ handleSubmit, hasError }) {
+  return <Layout onSubmit={handleSubmit} hasError={hasError} />;
 }
 
 Login.propTypes = {
   handleSubmit: PropTypes.func.isRequired,
-  hasErrorAuth: PropTypes.bool.isRequired,
-  msgError: PropTypes.string.isRequired
+  hasError: PropTypes.string
```
Comment (henryzarza) [inline training-react/tic-tac-toe/src/app/screens/Login/index.js]: Change this prop type for `bool`

File: training-react/tic-tac-toe/src/app/screens/Login/index.js [hunk 4]
```diff
@@ -5,19 +5,17 @@ import PropTypes from 'prop-types';
...
 };
 
 const mapStateToProps = state => ({
```
Comment (henryzarza) [inline training-react/tic-tac-toe/src/app/screens/Login/index.js]: You could do destructuring

File: training-react/tic-tac-toe/src/app/screens/Login/layout.js [hunk 5]
```diff
@@ -42,8 +42,7 @@ function Layout({ handleSubmit, hasError, msgError }) {
 
 Layout.propTypes = {
   handleSubmit: PropTypes.func.isRequired,
-  hasError: PropTypes.bool.isRequired,
-  msgError: PropTypes.string.isRequired
+  hasError: PropTypes.string
```
Comment (henryzarza) [inline training-react/tic-tac-toe/src/app/screens/Login/layout.js]: Change the name for something more descriptive

File: training-react/tic-tac-toe/src/redux/game/actions.js [hunk 6]
```diff
@@ -1,18 +1,16 @@
...
-  JUMP_TO_STEP: 'JUMP_TO_STEP',
-  CLICK_TO_SQUARE: 'CLICK_TO_SQUARE'
-};
+export const actionsTypes = createTypes(['JUMP_TO_STEP', 'CLICK_SQUARE'], '@@TIC_TAC_TOE');
 
 const actionCreators = {
   jumpToStep: step => ({
     type: actionsTypes.JUMP_TO_STEP,
-    step
+    target: 'game',
```
Comment (henryzarza) [inline training-react/tic-tac-toe/src/redux/game/actions.js]: Assign `game` target value to `const`

File: training-react/tic-tac-toe/src/redux/game/actions.js [hunk 7]
```diff
@@ -25,7 +23,8 @@ const actionCreators = {
     squares[i] = xIsNext ? 'X' : 'O';
 
     dispatch({
-      type: actionsTypes.CLICK_TO_SQUARE,
+      type: actionsTypes.CLICK_SQUARE,
+      target: 'game',
```
Comment (henryzarza) [inline training-react/tic-tac-toe/src/redux/game/actions.js]: The same here

File: training-react/tic-tac-toe/src/redux/game/reducer.js [hunk 8]
```diff
@@ -6,17 +8,21 @@ const initialState = {
...
-        stepNumber: action.step,
-        xIsNext: action.step % 2 === 0
-      };
-    case actionsTypes.CLICK_TO_SQUARE:
-      return { ...state, ...action.payload };
-    default:
-      return state;
-  }
-}
+const reducerEffects = {
```
Comment (henryzarza) [inline training-react/tic-tac-toe/src/redux/game/reducer.js:11]: Good approach 👏

File: training-react/tic-tac-toe/src/redux/login/actions.js [hunk 9]
```diff
@@ -1,44 +1,64 @@
...
-    let token;
-    const params = { username: userData.email, password: userData.password };
-    const response = await loginService.authUser(params);
-    if (response.ok) {
-      token = response.data.id;
-      localStorage.setItem('userToken', token);
-      localStorage.setItem('userId', response.data.userId);
+  authUser: userData => ({
+    type: actionsTypes.AUTH_USER,
+    target: target.AUTH,
```
Comment (henryzarza) [inline training-react/tic-tac-toe/src/redux/login/actions.js]: Change this for `IS_AUTH `

File: training-react/tic-tac-toe/src/redux/login/actions.js [hunk 10]
```diff
@@ -1,44 +1,64 @@
...
+          type: actionsTypes.SET_AUTH_STATE,
+          target: target.IS_AUTH,
+          payload: !!token
+        });
+      })
+    ]
+  }),
+  logout: () => ({
+    type: actionsTypes.AUTH_USER,
+    target: target.AUTH,
```
Comment (henryzarza) [inline training-react/tic-tac-toe/src/redux/login/actions.js]: The same here

**Review by henryzarza — CHANGES_REQUESTED** [review]
Good job, only a few comments.

**Review by henryzarza — DISMISSED** [review]

File: training-react/tic-tac-toe/package.json (hunk 2 above)
Comment (wfolini) [inline training-react/tic-tac-toe/package.json]: U aren't using this. Delete it

File: training-react/tic-tac-toe/src/redux/profile/reducer.js [hunk 11]
```diff
@@ -1,56 +1,41 @@
+import { completeState, createReducer, completeReducer, onReadValue } from 'redux-recompose';
+
 import { actionsTypes } from './actions';
 
-const initialState = {
-  info: {
+const initialStateDescription = {
+  getUser: null,
+  updateUser: null,
```
Comment (wfolini) [inline training-react/tic-tac-toe/src/redux/profile/reducer.js]: U don't need to save `getUser` and `updateUser`, Delete it

File: training-react/tic-tac-toe/src/redux/profile/actions.js [hunk 12]
```diff
@@ -1,44 +1,62 @@
...
+          type: actionsTypes.IS_LOADED_PROFILE,
+          target: target.IS_LOADED_PROFILE,
+          payload: true
+        });
+      })
+    ]
+  }),
+  updateUser: userData => ({
+    type: actionsTypes.UPDATE_USER,
+    target: target.UPDATE_USER,
```
Comment (wfolini) [inline training-react/tic-tac-toe/src/redux/profile/actions.js]: U can delete this target

File: training-react/tic-tac-toe/src/redux/profile/actions.js [hunk 13]
```diff
@@ -1,44 +1,62 @@
...
-        surname: response.data.surname,
-        username: response.data.username,
-        address: response.data.address
-      }
-    });
-    dispatch({ type: actionsTypes.LOADING, payload: true });
-  },
+  getUser: () => ({
+    type: actionsTypes.GET_USER,
+    target: target.GET_USER,
```
Comment (wfolini) [inline training-react/tic-tac-toe/src/redux/profile/actions.js]: Also this

File: training-react/tic-tac-toe/src/services/profileService.js [hunk 14]
```diff
@@ -1,8 +1,14 @@
 import api from '../config/api';
 
 const profileService = {
-  getUser: id => api.get(`/Users/${id}`),
-  updateUser: (id, params) => api.patch(`/Users/${id}`, params)
+  getUser: () => {
+    const idUSer = localStorage.getItem('userId');
```
Comment (wfolini) [inline training-react/tic-tac-toe/src/services/profileService.js]: U don't need to save this user id in the local storage. Save it in the redux store, and then, u can read it from a thunk action with `getState`. Check this https://github.com/reduxjs/redux-thunk#motivation

File: training-react/tic-tac-toe/src/services/profileService.js [hunk 15]
```diff
@@ -1,8 +1,14 @@
...
+    return api.get(`/Users/${idUSer}`);
+  },
+  updateUser: params => {
+    const idUSer = localStorage.getItem('userId');
```
Comment (wfolini) [inline training-react/tic-tac-toe/src/services/profileService.js]: Same here

**Review by wfolini — CHANGES_REQUESTED** [review]
Excellent! Good work! Some comments and go!

File: training-react/tic-tac-toe/src/redux/profile/actions.js (hunk 12 above)
Comment (andyOlarte514) [reply training-react/tic-tac-toe/src/redux/profile/actions.js]: if I remove the target, this happens:

image

File: training-react/tic-tac-toe/src/redux/profile/actions.js (hunk 13 above)
Comment (andyOlarte514) [reply training-react/tic-tac-toe/src/redux/profile/actions.js]: same

File: training-react/tic-tac-toe/src/redux/login/actions.js [hunk 16]
```diff
@@ -19,14 +20,20 @@ const actionCreators = {
     injections: [
       withPostSuccess((dispatch, response) => {
         const token = response.data.id;
+        const idUser = response.data.userId;
```
Comment (wfolini) [inline training-react/tic-tac-toe/src/redux/login/actions.js:23]: `const { token, userId }  = response.data;`

**Review by wfolini — APPROVED** [review]
Excellent!

**Pull request merged**